src/GraphQL/Instrumentation/GraphQLTelemetryProvider.cs

```diff
@@ -0,0 +1,183 @@
+#if NET5_0_OR_GREATER
+
+using System.Diagnostics;
+using GraphQL.DI;
+using GraphQL.Execution;
+using GraphQL.Types;
+using GraphQL.Validation;
+using GraphQLParser.AST;
+
+namespace GraphQL.Instrumentation;
+
+/// <summary>
+/// Provides telemetry through the <see cref="Activity">System.Diagnostics.Activity API</see> as an <see cref="IConfigureExecution"/> implementation.
+/// Derive from this class to add additional telemetry.
+/// <para>
+/// To use a derived class, call <see cref="GraphQLBuilderExtensions.ConfigureExecution{TConfigureExecution}(IGraphQLBuilder)"/> with the type of your derived class.
+/// </para>
+/// </summary>
+public class GraphQLTelemetryProvider : IConfigureExecution
+{
+    private readonly GraphQLTelemetryOptions _telemetryOptions;
+    private const string ACTIVITY_OPERATION_NAME = "graphql";
+
+    /// <summary>
+    /// Returns an <see cref="System.Diagnostics.ActivitySource"/> instance to be used for GraphQL.NET telemetry.
+    /// </summary>
+    protected static ActivitySource ActivitySource { get; }
+
+    /// <summary>
+    /// Gets the source name used for telemetry.
+    /// </summary>
+    public static string SourceName => "GraphQL";
+
+    static GraphQLTelemetryProvider()
+    {
+        ActivitySource = new ActivitySource(SourceName, typeof(GraphQLTelemetryProvider).Assembly.GetNuGetVersion());
+    }
+
+    /// <summary>
+    /// Initializes a new instance with the specified options.
+    /// </summary>
+    public GraphQLTelemetryProvider(GraphQLTelemetryOptions options)
+    {
+        _telemetryOptions = options;
+    }
+
+    /// <inheritdoc/>
+    public virtual float SortOrder => GraphQLBuilderExtensions.SORT_ORDER_CONFIGURATION;
+
+    /// <inheritdoc/>
+    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        if (!_telemetryOptions.Filter(options))
+            return await next(options).ConfigureAwait(false);
```

Comment:
I think the filter in this place is missing an important ability to filter on operation type because it's not available yet. Another option will be creating the activity without starting it (`Activity.CreateActivity`) and moving the filter into `TelemetryListener`, where the activity will be started based on the filter result (`activity.Start`). However, it has the drawback of unexpected order of delegates execution: the `EnrichWithExecutionOptions` will be called before `Filter`.

Reply:
Moving the `activity.Start` to a different point in time also affects proper date/time stamps.  Hence, assessment of the telemetry would not reflect the time spent decoding the request.

Reply:
You can set the start time immediately after creation or capture the creation time and set it before the start with [SetStartTime](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.setstarttime?view=net-7.0).

Reply:
what about sub-activities?  Would they be tracked/discarded correctly if the activity has not been started?  So for example, if there were activities recorded within the parser engine, which normally would automatically be a sub-activity -- would these be tracked/discarded properly if the activity was started late or never started?

Reply:
i'm thinking no.  not that we have any sub-activities yet, but just thinking forwards...

Reply:
Not sure. This is from [Start](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.start?view=net-7.0#remarks) method docs

> Calling the Start method does the following:
> 
> Sets [Parent](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.parent?view=net-7.0#system-diagnostics-activity-parent) to hold [Current](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.current?view=net-7.0#system-diagnostics-activity-current).
> Sets [Current](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.current?view=net-7.0#system-diagnostics-activity-current) to this [Activity](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity?view=net-7.0).
> If [StartTimeUtc](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.starttimeutc?view=net-7.0#system-diagnostics-activity-starttimeutc) was not set previously, sets it to [DateTime.UtcNow](https://learn.microsoft.com/en-us/dotnet/api/system.datetime.utcnow?view=net-7.0#system-datetime-utcnow).
> Generates a unique [Id](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.id?view=net-7.0#system-diagnostics-activity-id) for this activity.

So, I believe any activity started between `Create` and `Start` won't be a child of this activity, because `Current` was not replaced yet.

Reply:
Right, and it also would not be skipped.

Reply:
Then looks good to me! Thank you very much!

Reply:
> I think the filter in this place is missing an important ability to filter on operation type because it's not available yet. 

With a great desire, you can calculate operation type yourself, although I think that more often the user filter will use some unique information for each case (user IP, _trace_enabled_ headers/flags, etc).